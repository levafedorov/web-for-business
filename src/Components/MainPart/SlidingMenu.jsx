import React, {Component} from 'react';
import Positioner from '../Page/Positioner';
import './SlidingMenu.scss'
import slidingMenuHOC from './slidingMenuHOC';
import PropTypes from 'prop-types';

 class SlidingMenu extends Component{
    
   state = {
         targetPosition:0,
         targetWidth:null,
   }
  


   animTarget = (ev) => {
     
     const target = ev.target;
     const allAs = this.menuRef.querySelectorAll('a');
     const animElem = this.menuRef.querySelector('.anim-elem');
     let curIndex = [].findIndex.call(allAs, item => item === target);
     const workingArr = [...allAs].slice(0, curIndex);
     let targetPosition = 0; 
     
     allAs.forEach(item => item.classList.remove('active'));

     this.setClassNameToItem(target);

     animElem.classList.add('active');

     workingArr.forEach(item => targetPosition += item.offsetWidth);
     this.setState({targetPosition,
                    targetWidth:target.offsetWidth });

   }  
    
   setClassNameToItem = (elem, callback) => {
    setTimeout(() => {
      elem.classList.add('active');
      if(callback) callback();
      }, 300);
   }


   elemTransEnded = (ev) => {
    ev.target.classList.remove('active');
   }



    render(){
        
        const {targetPosition, targetWidth} = this.state;
        const {items, position} = this.props;
        
        const targetStyle = {transform:`translateX(${targetPosition}px)`,
                             width:targetWidth + 'px' }

        return <nav className="main-menu sliding" style={position}>
           <Positioner>
            <ul className="menu-list" ref={el => this.menuRef = el}>
            <div className="anim-elem" style={targetStyle} onTransitionEnd={this.elemTransEnded}></div>
    { Object.keys(items).map((item, i)=> <li key={item}> 
    <a href={items[item]} className={`link-${i} ${i === 0 ? 'active' : null}`} onClick={this.animTarget}>{item}</a></li>) }
            </ul>
           </Positioner>
        </nav>
    }
}


SlidingMenu.defaultProps = {
  items: {first:'first', second:'second', third:'third'},
}


SlidingMenu.propTypes = {
    items: PropTypes.objectOf(PropTypes.string),
    position: PropTypes.object
}





export const SlidingMenuTest = SlidingMenu;
export const SlidingMenuImpl = slidingMenuHOC(SlidingMenu);