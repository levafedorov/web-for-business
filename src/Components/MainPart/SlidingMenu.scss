@mixin flex($just:center, $al:center){
    display: flex;
      justify-content: $just; 
      align-items: $al;
   }
 
   @mixin flex-self($just:center, $al:center){
      justify-self: $just; 
      align-self: $al;
   }
 
 
   @mixin myriad($size:1em, $height:1em, $weight:normal){
     font-family: 'Myriad Pro Regular', 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
     font-size: $size;
     line-height: $height;
     font-weight: $weight;
 }
 
 @mixin baskerville($size:1em, $height:1em, $weight:normal){
    font-family: 'Baskerville Regular', 'Times New Roman', Times, serif;
    font-size: $size;
    line-height: $height;
    font-weight: $weight;
 }
 
 @mixin athelas($size:1em, $height:1em, $weight:normal){
    font-family: 'Athelas Regular', Georgia, 'Times New Roman', Times, serif;
    font-size: $size;
    line-height: $height;
    font-weight: $weight;
 }
 
 @mixin quatro($size:1em, $height:1em, $weight:normal){
    font-family: 'Quatro Regular' 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
    font-size: $size;
    line-height: $height;
    font-weight: $weight;
 }

 $white: white;
$black: black;
$dark-gray: #898d8e;
$light-gary: #b7bcbe;
$orange: #d7794a;
$whiten-gray:#e8e6e6;
$swampy: #1f3138;


@mixin decorator($bg:initial, $bord:none){
   background-color: $bg;
   border: $bord;
}


@mixin bg-image($bg){
  background-image: url($bg);
  background-repeat: no-repeat;
  background-size: cover;
}

 

   .sliding-menu-place{
      height: 78px;
   }
   .main-menu{
      z-index:1000;
      width: 100%;
      position:relative;
       @include flex();
       height: 50px;
       .positioner{
         height: 100%;
           @include flex(flex-start, center);
       }
       .menu-list{
           @include flex(flex-start, center);
           flex-basis: 785px;
           height: 100%;
           position: relative;
           li {
               position: relative;
               height: 100%;
               a {
                position: relative;
                padding: 0px 25px;
                @include flex();  
                height:100%;
                z-index: 3;
              }

           }
           .anim-elem{
            position:absolute;
            z-index: 2;
            width: 100px;
            height: 100%;
            transition: all 0.3s ease;
         }
       }
   }


   .main-menu.compact{
      height: 200px;
      flex-direction: column;
      transition: transform 0.3s ease;
      .menu-list{
         flex-direction: column;
         width: 100%;
        li{
           position: relative;
           display: block;
           width: 100%;
           &:before{
              width:0;
              z-index: 2;
              transition: all 0.3s ease;
              position: absolute;
              height: 100%;
              background-color: black;
              content: '';
           }
           &:hover:before{
              width: 100%;
           }
           & > a {
              position: relative;
              z-index: 3;
           }
        }
      }
      transform: translateY(-178px);

      .menu-down-btn{
         @include decorator();
         outline: none;
         margin-bottom:5px;
      }
      &:hover{
         transform: translateY(0);
      }
   }




@keyframes moving-target{

}





   @media screen and (max-width: 800px){
      .sliding-menu-place{
         height: 50px;
      }

      .main-part.default .main-menu .menu-list{ 
        
         & > li a{
            padding: 0 8px;
         }
      } 
   }



    