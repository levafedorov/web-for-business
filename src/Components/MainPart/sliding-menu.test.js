import React from 'react';
import {SlidingMenuTest} from './SlidingMenu';
import {shallow, mount} from 'enzyme';

const props = {items: {home: 'home', about: 'about', services: 'services'}}





describe('SlidingMenu component', () => {
    const setUp = (props) => mount(<SlidingMenuTest {...props}/>);
    let component;
    let instance;
    beforeEach(() => {
        component = setUp(props);
        instance = component.instance();
    });

    it("Should components prop's to be working", () => {
        component = setUp();
        expect(component).toMatchSnapshot();
        component = setUp(props);
        expect(component).toMatchSnapshot();
    });

    describe('Component methods', () => {
        test('should add class active when clicked', () => {
             let listItem = component.find('.link-2');
             instance.setClassNameToItem = jest.fn();
             listItem.simulate('click');
             expect(instance.setClassNameToItem).toHaveBeenCalledTimes(1);              
        });
         
            
});

});
