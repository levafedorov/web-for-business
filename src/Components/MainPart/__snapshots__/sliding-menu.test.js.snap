// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlidingMenu component Should components prop's to be working 1`] = `
<SlidingMenu
  items={
    Object {
      "first": "first",
      "second": "second",
      "third": "third",
    }
  }
>
  <nav
    className="main-menu sliding"
  >
    <Positioner>
      <div
        className="positioner"
      >
        <ul
          className="menu-list"
        >
          <div
            className="anim-elem"
            onTransitionEnd={[Function]}
            style={
              Object {
                "transform": "translateX(0px)",
                "width": "nullpx",
              }
            }
          />
          <li
            key="first"
          >
            <a
              className="link-0 active"
              href="first"
              onClick={[Function]}
            >
              first
            </a>
          </li>
          <li
            key="second"
          >
            <a
              className="link-1 null"
              href="second"
              onClick={[Function]}
            >
              second
            </a>
          </li>
          <li
            key="third"
          >
            <a
              className="link-2 null"
              href="third"
              onClick={[Function]}
            >
              third
            </a>
          </li>
        </ul>
      </div>
    </Positioner>
  </nav>
</SlidingMenu>
`;

exports[`SlidingMenu component Should components prop's to be working 2`] = `
<SlidingMenu
  items={
    Object {
      "about": "about",
      "home": "home",
      "services": "services",
    }
  }
>
  <nav
    className="main-menu sliding"
  >
    <Positioner>
      <div
        className="positioner"
      >
        <ul
          className="menu-list"
        >
          <div
            className="anim-elem"
            onTransitionEnd={[Function]}
            style={
              Object {
                "transform": "translateX(0px)",
                "width": "nullpx",
              }
            }
          />
          <li
            key="home"
          >
            <a
              className="link-0 active"
              href="home"
              onClick={[Function]}
            >
              home
            </a>
          </li>
          <li
            key="about"
          >
            <a
              className="link-1 null"
              href="about"
              onClick={[Function]}
            >
              about
            </a>
          </li>
          <li
            key="services"
          >
            <a
              className="link-2 null"
              href="services"
              onClick={[Function]}
            >
              services
            </a>
          </li>
        </ul>
      </div>
    </Positioner>
  </nav>
</SlidingMenu>
`;
